_posts/2016-04-01-rust-via-its-core-values.markdown

```diff
@@ -0,0 +1,342 @@
+
+Thanks to memory safety, we now understand some interesting things about Rust
+
+- Values can only have one owner
+- Values are destroyed when they leave scope
+
+Let's try to do something useful in Rust. Or as useful as can fit in this post. I want a function that tells me if two strings are the same length.
+
+First, functions. We declare them just like our main function:
+
+{% rp_highlight rust %}
+fn same_length() {
+}
+
+fn main() {
+  same_length();
+}
+{% endrp_highlight %}
+
+Our same_length function will need to accept two parameters, a source string and a other string to compare against.
+
+{% rp_highlight rust %}
+fn same_length(s1, s2) {
+}
+
+fn main() {
+  let source = String::from("Hello");
+  let other = String::from("Hi");
+
+  println!("{}", same_length(source, other));
+}
+//<anon>:1:18: 1:19 error: expected one of `:` or `@`, found `,`
+{% endrp_highlight %}
+
+Rust loves explicitness. So we can't declare a function without stating what kind of data we're going to pass into it. Rust uses strong, static typing in its function signatures. This way the compiler can ensure we're using our functions correctly, preventing exploding programs. Explicit typing also lets us easily see what a function accepts. Our function only accepts Strings, so we declare that.
+
+{% rp_highlight rust %}
+fn same_length(s1: String, s2: String) {
+}
+
+fn main() {
+  let source = String::from("Hello");
+  let other = String::from("Hi");
+
+  println!("{}", same_length(source, other));
+}
+//<anon>:8:18: 8:44 error: the trait `core::fmt::Display` is not implemented for the type `()` [E0277]
+{% endrp_highlight %}
+
+Most of Rust's compiler messages are helpful. This one, maybe not so much. What it's telling us is that our function is returning nothing, `()` and that can't print nothing. So our function should return...something. A boolean seems appropriate. Let's just return false for now.
+
+{% rp_highlight rust %}
+fn same_length(s1: String, s2: String) {
+  false
+}
+
+fn main() {
+  let source = String::from("Hello");
+  let other = String::from("Hi");
+
+  println!("{}", same_length(source, other));
+}
+//<anon>:2:3: 2:8 error: mismatched types:
+//expected `()`,
+//found `bool`
+{% endrp_highlight %}
+
+Explicitness again. Not only do functions have to declare what they accept, they have to declare the type of the object that they return. We return a `bool`
+
+{% rp_highlight rust %}
+#[allow(unused_variables)]
+fn same_length(s1: String, s2: String) -> bool {
+  false
+}
+
+fn main() {
+  let source = String::from("Hello");
+  let other = String::from("Hi");
+
+  println!("{}", same_length(source, other));
+}
+//false
+{% endrp_highlight %}
+
+Cool. That compiles. Let's do an actual comparison. Strings have a `len()` function that returns their length:
+
+{% rp_highlight rust %}
+fn same_length(s1: String, s2: String) -> bool {
+  s1.len() == s2.len()
```

Comment:
All good, no worries. 
I figured, since it's just comparing the lengths, it doesn't matter the unit, as long as it's the same unit for both.
`s1.len() == s2.len()`
vs.
`s1.chars().count() == s2.chars().count()`

[That](https://github.com/IanWhitney/designisrefactoring/pull/19/files/76c0133e91d00a52e3a4aee6d40d42eb3dea862e#r58469092) is actually good information for I didn't know about the distinction(looking it up currently). Thanks!